crates/iota-mainnet-unlocks/src/store.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeMap, fs, path::PathBuf};
+
+use anyhow::{Context, Result};
+use chrono::{DateTime, Utc};
+use csv::ReaderBuilder;
+use serde::{Deserialize, Serialize};
+
+/// File name of the mainnet unlock data.
+pub const INPUT_FILE: &str = "aggregated_mainnet_unlocks.csv";
+
+/// Represents a single entry in the store.
+/// It defines how many tokens still remain locked at a specific point in time.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct StillLockedEntry {
+    /// UTC timestamp at which the tokens are still locked.
+    pub timestamp: DateTime<Utc>,
+    /// Total locked amount (nano-units) still locked at the timestamp.
+    pub amount_still_locked: u64,
+}
+
+/// In-memory store holding the aggregated token unlock data.
+#[derive(Debug, Clone)]
+pub struct MainnetUnlocksStore {
+    // Each entry represents the total number of tokens still locked at the specific point in time.
+    entries: BTreeMap<DateTime<Utc>, StillLockedEntry>,
+}
+
+impl MainnetUnlocksStore {
+    /// Creates a new store with the aggregated unlock data for mainnet.
+    /// Loads the aggregated token unlock data from the given JSON file at the
+    /// crate root.
+    pub fn new() -> Result<Self> {
+        let crate_dir = env!("CARGO_MANIFEST_DIR");
+        let path = PathBuf::from(crate_dir).join("data").join(INPUT_FILE);
+
+        let data = fs::read_to_string(&path)
+            .with_context(|| format!("could not read locked supply file: {:?}", path))?;
+
+        Self::from_csv_str(&data)
+    }
+
+    /// Parses the given CSV string into a `MainnetUnlocksStore`.
+    fn from_csv_str(csv_str: &str) -> Result<Self> {
+        let mut rdr = ReaderBuilder::new()
+            .has_headers(true)
+            .from_reader(csv_str.as_bytes());
+
+        let mut map = BTreeMap::new();
+
+        for result in rdr.deserialize() {
+            let entry: StillLockedEntry =
+                result.context("failed to deserialize CSV row into StillLockedEntry")?;
+
+            if map.contains_key(&entry.timestamp) {
```

Comment:
Not very important but FYI `insert` returns whether the entry existed or not, so `contains_key` is a bit redundant

Reply:
Good point, made use of that `Option` in https://github.com/iotaledger/iota/pull/6123/commits/1a16ba645060ca55ffe5edd134c95a8cc66570ac